markdown/docs/addon-author-guide.md

```diff
@@ -0,0 +1,321 @@
+do as little work as possible during the evaluation of your addon's
+JavaScript files. Doing upfront work can have a negative impact on load
+times. If you can defer doing work (such as feature detection) until the
+last possible moment, you can improve the performance of applications
+using your addon. Adding FastBoot compatibility is a good opportunity to
+make these improvements to your addon.
+
+Here's a concrete example. Imagine our addon has a component that needs
+to listen to `resize` events on the window. We set up a single listener in
+our component file that's shared across all instances of the component:
+
+```js
+// addon/components/resizable-component.js
+import Component from "ember-component";
+import $ from "jquery";
+
+let componentsToNotify = [];
+$(window).on('resize', () => {
+  componentsToNotify.forEach(c => c.windowDidResize());
+});
+
+export default Component.extend({
+  init() {
+    componentsToNotify.push(this);
+  },
+
+  windowDidResize() {
+    // ... do some work
+  },
+
+  willDestroy() {
+    componentsToNotify.removeObject(this);
+  }
+});
+```
+
+In this approach, we set up an event listener on the window even if the
+app author never uses our component. It also means that our addon won't
+work in FastBoot, because it tries to access the DOM while defining the
+component.
+
+Instead, we can move this setup into the component's
+`didInsertElement()` hook, which is only invoked once the component is
+inserted into the DOM (i.e., it's only invoked in the browser
+and never in Node).
+
+```js
+// addon/components/resizable-component.js
+import Component from "ember-component";
+import $ from "jquery";
+
+let componentsToNotify = [];
+let didSetupListener = false;
+
+function setupListener() {
+  didSetupListener = true;
+  $(window).on('resize', () => {
+    componentsToNotify.forEach(c => c.windowDidResize());
+  });
+}
+
+export default Component.extend({
+  didInsertElement() {
+    if (!didSetupListener) { setupListener(); }
+    componentsToNotify.push(this);
+  },
+
+  windowDidResize() {
+    // ... do some work
+  },
+
+  willDestroy() {
+    componentsToNotify.removeObject(this);
+  }
+});
+```
+
+This lazy approach moves setup to the first time the component is used,
+improving boot time for routes in the application that don't use this
+component. It also means the component is FastBoot-compatible, because
+all of the code that touches the DOM is contained in the
+`didInsertElement()` hook, which is not invoked in FastBoot.
+
+### Third-Party Dependencies
+
+Many Ember addons wrap other JavaScript libraries in a way that makes
+them easier to use from an Ember app. For example, the
+[ivy-codemirror](https://github.com/IvyApp/ivy-codemirror) addon wraps
+the [CodeMirror code editor](https://codemirror.net/) in a component
+that makes it easy to drop into an Ember app.
+
+Sometimes the library your addon uses is itself incompatible with
```

Comment:
@locks This is technically grammatically correct, but I can see why it reads funny. Maybe this is clearer:

"Sometimes the library your addon wraps is itself incompatible with"